src/power_grid_model/core/serialization.py

```diff
@@ -80,7 +85,33 @@ def load(self) -> Dataset:
             A tuple containing the deserialized dataset in Power grid model input format and the type of the dataset.
         """
         pgc.deserializer_parse_to_buffer(self._deserializer)
-        return self._dataset.get_data()
+
+        filtered_data = self._dataset.get_data()
+
+        def _filter_attributes(attributes):
+            keys_to_remove = []
+            for attr, array in attributes.items():
+                if not isinstance(array, np.ndarray):
+                    continue
+                if is_nan_or_equivalent(array):
+                    keys_to_remove.append(attr)
+            for key in keys_to_remove:
+                del attributes[key]
+
+        if isinstance(self._data_filter, ComponentAttributeFilterOptions):
+            if self._data_filter is ComponentAttributeFilterOptions.RELEVANT:
+                for _, attributes in filtered_data.items():
+                    _filter_attributes(attributes)
+        elif isinstance(self._data_filter, dict):
+            for component_type, attributes in filtered_data.items():
+                if component_type in self._data_filter:
```

Comment:
> what if a component type was requested via the self._data_filter but not present in the filtered_data? It should be present, just like it would be with the attributes

ignore

> what if a component type was present in filtered_data but not in the self._data_filter?

`_data_filter` major here: if not mentioned in the `_data_filter` we do not touch it